---
weight: 30
---

# RabbitMQ Mnesia 数据库异常处理

## Mnesia 数据库常见故障说明

RabbitMQ 使用 Mnesia 数据库存储队列、交换器、绑定等信息。导致 Mnesia 失败的常见原因有两类：

* MNESIA_BASE 目录权限问题，用户对该目录的写权限不足
* Mnesia 读取 table 失败

## 权限问题

当您遇到 Mnesia 数据库权限问题时，请为 MNESIA_BASE 目录配置好当前用户的写权限即可解决。

## Mnesia 读取 table 失败

Mnesia 会基于机器的主机名来创建对应的数据库 schema，因此在主机名发生变更时，Mnesia 就无法载入旧的 schema；同样，如果对 rabbit@hostname 目录进行了重命名，那么 Mnesia 也将无法找到旧的数据库文件，此时会重新创建 rabbit@hostname 文件夹并启动数据库。

导致 Mnesia 读取失败可以参考以下步骤进行修复：

1. 首先判断节点的主机名称或 rabbit@hostname 目录名称是否有更新，如果有更新可以重命名，重命名规范为 “rabbit@hostname”，重命名后即可看到旧的数据库文件。

2. 如果未发生更新，在集群模式下，可能会因为启动时各副本间无法成功连接，可以根据日志排查副本间的通信问题。

3. 上述仍不能恢复数据，此时可以选择对当前集群进行配置备份，重新部署新的节点加入原有集群，并导入备份好的配置。

## 其他与 Mnesia 数据库有关的故障

如果在 RabbitMQ 重启后，无法正常组件新集群，日志中会如下类型的错误:

```
[warning] <0.273.0> Error while waiting for Mnesia tables: {timeout_waiting_for_tables,['rabbit@e2e-rabbitmq-server-2.e2e-rabbitmq-nodes.local-midautons','rabbit@e2e-rabbitmq-server-1.e2e-rabbitmq-nodes.local-midautons','rabbit@e2e-rabbitmq-server-0.e2e-rabbitmq-nodes.local-midautons'],[rabbit_durable_queue]}
 [info] <0.273.0> Waiting for Mnesia tables for 30000 ms, 7 retries left
```

此问题是因为在 Mnesia 数据库中存储有集群节点的信息，重启 RabbitMQ 时没有正确的清理这些状态信息，由于 RabbitMQ 的 Pod 启动策略为顺序启动，这样会导致先启动的 Pod 会一直等待其他 Pod 启动完成，形成死循环。建议处理办法为强制删除集群状态并重新启动，您可以在 kubectl 中执行以下命令：

```
kubectl -n {命名空间名称} exec {实例名称} -- rabbitmqctl force_boot
```